### PR TITLE
Fix SPDY proxy authentication with percent escaped characters

### LINKED ISSUE #791: kubectl exec & port forward TLS handshake error when https_proxy environment variable contains special characters
**kubectl version**

- **Client Version**: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.12", GitCommit:"a8b52209ee172232b6db7a6e0ce2adc77458829f", GitTreeState:"clean", BuildDate:"2019-10-15T12:12:15Z", GoVersion:"go1.11.13", Compiler:"gc", Platform:"windows/amd64"}
- **Server Version**: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.12", GitCommit:"524c3a1238422529d62f8e49506df658fa9c8b8c", GitTreeState:"clean", BuildDate:"2019-11-14T05:26:24Z", GoVersion:"go1.11.13", Compiler:"gc", Platform:"linux/amd64"}

**Issue**

When running the following behind a corporate proxy:

kubectl exec -it grafana-8665c55f6b-k7jpb ls -n monitoring

I'm seeing:

error: error sending request: Post https://k8s-fqdn:443/api/v1/namespaces/monitoring/pods/grafana-8665c55f6b-k7jpb/exec?command=ls&container=grafana&container=grafana&stdin=true&stdout=true&tty=true: tls: first record does not look like a TLS handshake

The same issue also happens with port forwarding e.g.:

kubectl -n monitoring port-forward prometheus-server-657c6d8f59-s7mjz 9090

error: error upgrading connection: error sending request: Post https://k8s-fqdn:443/api/v1/namespaces/monitoring/pods/prometheus-server-657c6d8f59-s7mjz/portforward: tls: first record does not look like a TLS handshake

**How to reproduce**

If the password for the https_proxy environment variable contains a hash symbol (URL encoded this is %23), it doesn't work and we get a TLS handshake error. If the password for the https_proxy environment variable contains a dollar symbol (URL encoded this is %24) it works fine.

The following **does not** work and gives the TLS handshake error above:

http_proxy=http://username:password%23@proxyhost:proxy-port
https_proxy=http://username:password%23@proxyhost:proxy-port

The following **does** work:

http_proxy=http://username:password%24@proxyhost:proxy-port
https_proxy=http://username:password%24@proxyhost:proxy-port

and we see:

**kubectl exec -it grafana-8665c55f6b-k7jpb ls -n monitoring**
LICENSE    README.md  bin        public     tools
NOTICE.md  VERSION    conf       scripts

OR

**kubectl -n monitoring port-forward prometheus-server-657c6d8f59-s7mjz 9090**
Forwarding from 127.0.0.1:9090 -> 9090
Forwarding from [::1]:9090 -> 9090

as expected.

### PR DESCRIPTION
**Pull request by 3point2**
Percent escaped characters in the proxy URL are now correctly used unescaped when constructing the proxy auth header for SPDY connections.

#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
kubectl exec fails when using a proxy that requires authentication and the user's credentials include percent escaped characters. see kubernetes/kubectl#791 for details.

#### Which issue(s) this PR fixes:

Fixes kubernetes/kubectl#791

#### Special notes for your reviewer:
None

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by 3point2** [comment]
/priority P2
(carried over from kubernetes/kubectl#791)

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go [hunk 1]
```diff
@@ -184,12 +184,15 @@ func (s *SpdyRoundTripper) dialWithHttpProxy(req *http.Request, proxyURL *url.UR
...
 	//nolint:staticcheck // SA1019 ignore deprecated httputil.NewProxyClientConn
 	proxyClientConn := httputil.NewProxyClientConn(proxyDialConn, nil)
-	_, err = proxyClientConn.Do(&proxyReq)
+	response, err := proxyClientConn.Do(&proxyReq)
 	//nolint:staticcheck // SA1019 ignore deprecated httputil.ErrPersistEOF: it might be
 	// returned from the invocation of proxyClientConn.Do
 	if err != nil && err != httputil.ErrPersistEOF {
 		return nil, err
 	}
+	if response != nil && response.StatusCode != http.StatusOK {
```
Comment (aojea) [inline staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:193]: how is this related?

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go [hunk 2]
```diff
@@ -401,15 +411,26 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 			}
 
 			var expectedProxyAuth string
+			var expectedCredentials string
 			if testCase.proxyAuth != nil {
-				encodedCredentials := base64.StdEncoding.EncodeToString([]byte(testCase.proxyAuth.String()))
-				expectedProxyAuth = "Basic " + encodedCredentials
+				expectedPassword, _ := testCase.proxyAuth.Password()
```
Comment (aojea) [inline staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:416]: don't you have to check if the password is set?

Comment (aojea) [reply staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:416]: I didn't look too much , but it will be nice if we can reuse some of the golang std library methods instead of this manual parsing https://pkg.go.dev/net/http#Request.BasicAuth

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go (hunk 1 above)
Comment (aojea) [reply staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:193]: separate commits are better in this case , imagine we had to revert because of X reasons, we lost this change that is useful

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go (hunk 2 above)
Comment (aojea) [reply staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:416]: I was curious about the password because in the code above you check that is not empty to generate the credentials

> > 	proxyAuthUser := proxyURL.User.Username()
> 	proxyAuthPassword, proxyAuthPasswordSet := proxyURL.User.Password()
> 	credentials := proxyAuthUser
> 	if proxyAuthPasswordSet {
> 		credentials += ":" + proxyAuthPassword
> 	}

This is what we should do  https://cs.opensource.google/go/go/+/master:src/net/http/transport.go;l=844-856;drc=c81c0279827545b20f81797bb263ae696d9c235f 

in our `func (s *SpdyRoundTripper) proxyAuth(proxyURL *url.URL) string {`

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go [hunk 3]
```diff
@@ -294,7 +297,12 @@ func (s *SpdyRoundTripper) proxyAuth(proxyURL *url.URL) string {
 	if proxyURL == nil || proxyURL.User == nil {
 		return ""
 	}
-	credentials := proxyURL.User.String()
+	proxyAuthUser := proxyURL.User.Username()
+	proxyAuthPassword, proxyAuthPasswordSet := proxyURL.User.Password()
+	credentials := proxyAuthUser
+	if proxyAuthPasswordSet {
+		credentials += ":" + proxyAuthPassword
+	}
```
Comment (aojea) [inline staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:305]: I think that we should do the same as the std library https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/net/http/transport.go;l=846;drc=e822b1e26e20ef1c76672c0b77b0fd8a97a1fe84
 , something like 

```
		username := u.Username()
		password, _ := u.Password()
	credentials := fmt.Sprintf("%s:%s", username, password)
```

**Comment by aojea** [comment]
/ok-to-test

it lgtm, just some comments to split into 2 different commits and refactor to match the std library logic

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go (hunk 1 above)
Comment (3point2) [reply staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:193]: :+1:  done

**Review by 3point2 — COMMENTED** [review]
I've reworked the construction of the proxy auth header to match the [standard library code](https://cs.opensource.google/go/go/+/master:src/net/http/transport.go;l=844-856;drc=c81c0279827545b20f81797bb263ae696d9c235f) that you linked to and split into two commits @aojea, let me know what you think.

**Comment by aojea** [comment]
/lgtm
/assign @liggitt

Jordan, for context https://github.com/kubernetes/kubernetes/pull/113283#discussion_r1011005114 , this replaces the auth logic to match the std lib

**Comment by 3point2** [comment]
@liggitt just checking in as it's been a week - do you think you'll be able to find a bit of time to look at this PR please?

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go (hunk 1 above)
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:193]: limiting this to exactly 200 does not seem correct... change this to accept any 2xx status code, per https://www.rfc-editor.org/rfc/rfc7231#section-4.3.6:

>   Any 2xx (Successful) response indicates that the sender (and all
>   inbound proxies) will switch to tunnel mode immediately after the
>   blank line that concludes the successful response's header section;
>   data received after that blank line is from the server identified by
>   the request-target.  Any response other than a successful response
>   indicates that the tunnel has not yet been formed and that the
>   connection remains governed by HTTP.

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go [hunk 4]
```diff
@@ -400,16 +410,27 @@ func TestRoundTripAndNewConnection(t *testing.T) {
...
-			var expectedProxyAuth string
+			var auth string
+			var authHeader string
 			if testCase.proxyAuth != nil {
-				encodedCredentials := base64.StdEncoding.EncodeToString([]byte(testCase.proxyAuth.String()))
-				expectedProxyAuth = "Basic " + encodedCredentials
+				username := testCase.proxyAuth.Username()
+				password, _ := testCase.proxyAuth.Password()
+				auth = username + ":" + password
+				authHeader = "Basic " + base64.StdEncoding.EncodeToString([]byte(auth))
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:419]: keep the name of this `expectedProxyAuth` to make it clear which is the data in the testcase and which is the data we're expecting

both the old test code and new test code share a flaw of essentially duplicating the implementation, so if the implementation is wrong, our copy in the test will be wrong too, and the test will still pass

can we improve this to actually extract the username and password from the header using a stdlib method and see if the username and password match the testcase data? (example at https://go.dev/play/p/psQsaWVA4ZG)

**Review by liggitt — COMMENTED** [review]
Sorry, this got stuck in the crush of code freeze reviews. Did a sweep

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go (hunk 1 above)
Comment (3point2) [reply staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:193]: you're right, thanks for catching that

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go [hunk 5]
```diff
@@ -291,6 +291,16 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 			serverStatusCode:       http.StatusSwitchingProtocols,
 			shouldError:            false,
 		},
+		"proxied valid https, proxy auth with special char in password -> valid https": {
+			serverFunc:             httpsServerValidHostname(t),
+			proxyServerFunc:        httpsServerValidHostname(t),
+			proxyAuth:              url.UserPassword("proxyuser", "proxypasswd%"),
```
[the remark this answers is not in the document]
Comment (3point2) [reply staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go]: done

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go (hunk 4 above)
Comment (3point2) [reply staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go]: it's now expectedUsername and expectedPassword, thanks for the suggestion. although it doesn't get rid of all the duplication of logic, at least the base64 decoding is now being handled by stdlib.

**Review by 3point2 — COMMENTED** [review]
> Sorry, this got stuck in the crush of code freeze reviews

No problem, thanks for replying and reviewing so quickly! I've tried to address your comments, please let me know what you think.

**Comment by liggitt** [comment]
/lgtm
/approve

I don't feel strongly about this being included in the 1.26 milestone, but I'd be supportive of a cherry-pick to release branches. Not sure how the release team would want a low risk bugfix PR like that dispositioned.

**Comment by 3point2** [comment]
Thanks for the review!

Is there anything else I need to do to help this get merged, or is it just a matter of waiting until the 1.26 freeze ends?

**Comment by liggitt** [comment]
you're good, just waiting for master to open for 1.27

**Comment by dims** [comment]
see discussion above, punting to post code freeze.

/milestone v1.27
